tests/test_api.py

```diff
@@ -0,0 +1,1576 @@
+from gsql.backend import api_handler
+from googleapiclient import errors
+
+
+def test_init_success(monkeypatch, caplog):
+    def mock_build_success(*args, **kwargs):
+        return "mock_build_obj"
+
+    monkeypatch.setattr(api_handler, "build", mock_build_success)
+    instance = api_handler.ApiHandler("creds")
+    assert instance.creds == "creds"
+    assert instance.gdriveService == "mock_build_obj"
+    assert instance.sheetService == "mock_build_obj"
+
+
+def test_init_failure_error(monkeypatch, mocker, caplog):
+    def mock_build_failure(*args, **kwargs):
+        raise errors.Error("invalid cred error")
+
+    monkeypatch.setattr(api_handler, "build", mock_build_failure)
+    instance = api_handler.ApiHandler("creds")
+    assert instance.creds == "creds"
+    assert (
+        caplog.records[0].msg == "driver creation failed with error: invalid cred error"
+    )
+
+
+def test_init_failure_exception(monkeypatch, mocker, caplog):
+    def mock_build_failure(*args, **kwargs):
+        raise Exception("invalid cred exception")
+
+    monkeypatch.setattr(api_handler, "build", mock_build_failure)
+    instance = api_handler.ApiHandler("creds")
+    assert instance.creds == "creds"
+    assert (
+        caplog.records[0].msg
+        == "driver creation failed with error: invalid cred exception"
+    )
+
+
+def test_create_spreadsheet_success(monkeypatch, mocker, caplog):
+    class MockSpreadSheetService:
```

Comment:
Do not create this class *every time* inside a test case. 
Declare the class outside test cases and use it for every test case.


You can create a base Mock class and override the changing functions which in most cases will be `self.execute` and you can also change attributes such as `mock_spreadsheet` by `object.mock_spreadsheet = value`